### ISSUE TITLE
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory

### ISSUE THREAD

**Issue by akd3vs** [dyatko/arkit#7]
Hi! This project looks very interesting. I wanted to try it with a somewhat medium project of mine, and it throws the following error:

`arkit

<--- Last few GCs --->

[3367:0x2768af0]    53796 ms: Scavenge 1382.2 (1424.7) -> 1381.5 (1425.2) MB, 4.3 / 0.0 ms  (average mu = 0.300, current mu = 0.258) allocation failure 
[3367:0x2768af0]    53802 ms: Scavenge 1382.2 (1425.2) -> 1381.6 (1425.7) MB, 4.3 / 0.0 ms  (average mu = 0.300, current mu = 0.258) allocation failure 
[3367:0x2768af0]    53807 ms: Scavenge 1382.4 (1425.7) -> 1381.8 (1426.2) MB, 3.4 / 0.0 ms  (average mu = 0.300, current mu = 0.258) allocation failure 

<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x329731dcfc5d]
Security context: 0x1dd77731d971 <JSObject>
    1: next [0x1dd777319711](this=0x2a7e75a871c1 <MapIterator map = 0xe2347085bd1>)
    2: /* anonymous */ [0x10b6bb2d4ec1] [/app/frontend/node_modules/ts-morph/dist/utils/ArrayUtils.js:~59] [pc=0x329735b54ad8](this=0x337c958302c1 <JSFunction ArrayUtils (sfi = 0x10b6bb2d40d9)>,0x2a7e75a871c1 <MapIterator map = 0xe2347085bd1>)
    3: /...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory

Writing Node.js report to file: report.20190226.145847.3367.001.json
Node.js report completed
 1: 0x953b10 node::Abort() [node]
 2: 0x9547f4 node::OnFatalError(char const*, char const*) [node]
 3: 0xb32bee v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0xb32e24 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0xf32452  [node]
 6: 0xf32558 v8::internal::Heap::CheckIneffectiveMarkCompact(unsigned long, double) [node]
 7: 0xf3ec78 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [node]
 8: 0xf3f78b v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 9: 0xf424c1 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [node]
10: 0xf0c6f4 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [node]
11: 0x11c2b3e v8::internal::Runtime_AllocateInNewSpace(int, v8::internal::Object**, v8::internal::Isolate*) [node]
12: 0x329731dcfc5d 
Aborted (core dumped)`

would you know if there's something I could try to fix this? Also let me know if you need the report.

Cheers!

**Comment by dyatko** [comment]
@akd3vs I'm glad to help but need a little bit more details:
- Operation system
- Node version
- Exact command you're running
- arkit.json if any used

**Comment by akd3vs** [comment]
@dyatko Thank you! Sure thing, I'm sorry for not giving more details.

- Linux (Ubuntu, also tried with Fedora 29)
- 11.10.0
- Using this config: `https://github.com/dyatko/arkit/blob/master/test/react-dom/arkit.json` and just ran `arkit`

Some more details about my project: It's basically create react app with almost full structure (services, redux-saga, translations, containers, components, using react 16.8, bootstrap, etc. I'll be happy to provide more details if needed.

Thank you again and cheers!

**Comment by dyatko** [comment]
@akd3vs please try this config, which excludes node_modules folder

```json
{
  "$schema": "https://raw.githubusercontent.com/dyatko/arkit/master/schema.json",
  "excludePatterns": ["node_modules/**", "test/**", "tests/**", "**/*.test.*", "**/*.spec.*"],
  "components": [
    {
      "type": "Entry",
      "patterns": ["*.js"]
    },
    {
      "type": "Component",
      "patterns": ["src/**/*.js"],
      "excludePatterns": ["**/__tests__/**", "**/*-test.js"]
    }
  ],
  "output": {
    "path": "arkit.svg"
  }
}
```

**Comment by akd3vs** [comment]
@dyatko With the exact config it gave the same message. With the following it gave a JS error:

`{
  "$schema": "https://raw.githubusercontent.com/dyatko/arkit/master/schema.json",
  "excludePatterns": ["node_modules/**", "test/**", "tests/**", "**/*.test.*", "**/*.spec.*", "src/mdbreact/**", "src/svg/**", "build/**", "coverage/**", "cypress/**"],
  "components": [
    {
      "type": "Entry",
      "patterns": ["*.js", "*.jsx"]
    },
    {
      "type": "Component",
      "patterns": ["src/**/*.js", "src/**/*.jsx"],
      "excludePatterns": ["**/__tests__/**", "**/*-test.js", "**/*.spec.js", "**/__mocks__/**"]
    }
  ],
  "output": {
    "path": "arkit.svg"
  }
}
`

And the error:

`node_modules/ts-morph/dist/compiler/ast/base/name/DeclarationNamedNode.js:35
                    throw errors.getNotImplementedForSyntaxKindError(compilerNameNode.kind);
                    ^

Error: Not implemented feature for syntax kind 'ObjectBindingPattern'.
    at NotImplementedError.BaseError [as constructor] (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/errors/classes/BaseError.js:8:28)
    at new NotImplementedError (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/errors/classes/NotImplementedError.js:10:23)
    at Object.getNotImplementedForSyntaxKindError (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/errors/helpers.js:79:12)
    at ParameterDeclaration.DeclarationNamedNodeInternal.class_1.getNameNode (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/base/name/DeclarationNamedNode.js:35:34)
    at ParameterDeclaration.DeclarationNamedNodeInternal.class_1.getName (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/base/name/DeclarationNamedNode.js:42:33)
    at ParameterDeclaration.DeclarationNamedNodeInternal.class_1.getStructure (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/base/name/DeclarationNamedNode.js:57:28)
    at Object.callBaseGetStructure (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/callBaseGetStructure.js:9:51)
    at ParameterDeclaration.InitializerSetExpressionableNode.class_1.getStructure (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/base/initializer/InitializerSetExpressionableNode.js:50:43)
    at Object.callBaseGetStructure (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/callBaseGetStructure.js:9:51)
    at ParameterDeclaration.TypedNode.class_1.getStructure (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/base/TypedNode.js:85:43)
`

Sorry, I forgot that I had a dependency in a folder inside src. Also excluded other folders that could be big and I'm not interested in analyzing.

Let me know if you have another idea of what's wrong.

Cheers.

**Comment by dyatko** [comment]
@akd3vs seems like ts-morph fails on some ObjectBinding pattern.
If you run arkit as `LEVEL=debug arkit`, then we can figure out which source file causes this issue and, thus, the object binding itself.

If you provide object binding code sample, I can figure out some solution.

**Comment by dyatko** [comment]
@akd3vs I just added object binding test and it works well https://github.com/dyatko/arkit/commit/fd8f56e9bf0eb8a7227e32af22755ac5f8a57a1a

I suggest you try the latest arkit version

**Comment by akd3vs** [comment]
@dyatko Thank you for your suggestion. Just tried the new version and it throwed the same error, and with debug I saw that the file where I configure redux is the one causing this error, excluding it now yields the svg correctly. Would it be helpful to you if I send you the code? Right now my next move will be to create a new schema for a complex React project so it can accurate represent a hierarchy of pages.

Output:

```
[2019-03-05 14:40:58.121 +0000] DEBUG: src/store.js 11 statements
/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/base/name/DeclarationNamedNode.js:35
                    throw errors.getNotImplementedForSyntaxKindError(compilerNameNode.kind);
                    ^

Error: Not implemented feature for syntax kind 'ObjectBindingPattern'.
    at NotImplementedError.BaseError [as constructor] (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/errors/classes/BaseError.js:8:28)
    at new NotImplementedError (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/errors/classes/NotImplementedError.js:10:23)
    at Object.getNotImplementedForSyntaxKindError (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/errors/helpers.js:79:12)
    at ParameterDeclaration.DeclarationNamedNodeInternal.class_1.getNameNode (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/base/name/DeclarationNamedNode.js:35:34)
    at ParameterDeclaration.DeclarationNamedNodeInternal.class_1.getName (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/base/name/DeclarationNamedNode.js:42:33)
    at ParameterDeclaration.DeclarationNamedNodeInternal.class_1.getStructure (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/base/name/DeclarationNamedNode.js:57:28)
    at Object.callBaseGetStructure (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/callBaseGetStructure.js:9:51)
    at ParameterDeclaration.InitializerSetExpressionableNode.class_1.getStructure (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/base/initializer/InitializerSetExpressionableNode.js:50:43)
    at Object.callBaseGetStructure (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/callBaseGetStructure.js:9:51)
    at ParameterDeclaration.TypedNode.class_1.getStructure (/home/andre/WebstormProjects/frontend/node_modules/ts-morph/dist/compiler/ast/base/TypedNode.js:85:43)
andre@andre-workstation:~/WebstormProjects/frontend$ npm ls arkit
frontend@0.2.0 /home/andre/WebstormProjects/frontend
└── arkit@1.2.4
```

**Comment by dyatko** [comment]
@akd3vs Yes please, share this `store.js` file, then I will definitely reproduce and fix that.

If you don't want to share the complete file, you can try to remove some lines and run arkit again to find the statement it cannot parse.

**Comment by akd3vs** [comment]
@dyatko Here you go: https://gist.github.com/akd3vs/2fe1b9c004069d8b0030402a344975fa

unfortunately that's as far as I can disclosure. Haven't test removing lines but I will do it since it would be nice to include the redux structure in the diagram. If you find something let me know, if I find something I'll let you know as soon as I can!

Thank you for the help!

**Comment by dyatko** [comment]
It failed on this export
```js
export function injectAsyncSagas({ name, sagas, store }) {
  if (!store.asyncSagas.includes(name)) {
    sagas.forEach(store.runSaga);
    store.asyncSagas.push(name);
  }
}
```

Fixed in v1.2.7, please give it a try
